Fail to build test with googletest 1.11.0

Hi, I'm updating sdbus-cpp in [Debian](https://tracker.debian.org/pkg/sdbus-cpp). As Debian has bumped googletest to 1.11.0, I fail to build sdbus-cpp's test.

Some relevant log

```
/usr/src/googletest/googletest/include/gtest/gtest-printers.h:291:36: error: no matching function for call to ‘testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter::PrintValue(const std::chrono::duration<long int, std::ratio<1, 1000000> >&, std::nullptr_t)’
  291 |     T, decltype(Printer::PrintValue(std::declval<const T&>(), nullptr)),
      |                 ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/googletest/googletest/include/gtest/gtest-printers.h:212:15: note: candidate: ‘template<class T, class, class> static void testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter::PrintValue(const T&, std::ostream*)’
  212 |   static void PrintValue(const T& value, ::std::ostream* os) {
      |               ^~~~~~~~~~
/usr/src/googletest/googletest/include/gtest/gtest-printers.h:212:15: note:   template argument deduction/substitution failed:
/usr/src/googletest/googletest/include/gtest/gtest-printers.h:211:33: error: no match for ‘operator<<’ (operand types are ‘std::basic_ostream<char>’ and ‘const std::chrono::duration<long int, std::ratio<1, 1000000> >’)
  210 |             typename = decltype(std::declval<std::ostream&>()
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  211 |                                 << std::declval<const T&>())>
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /<<PKGBUILDDIR>>/tests/integrationtests/Defs.h:32,
                 from /<<PKGBUILDDIR>>/tests/integrationtests/DBusConnectionTests.cpp:28:

```

```
/usr/include/c++/11/ostream:750:5: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
In file included from /usr/src/googletest/googletest/include/gtest/gtest.h:55,
                 from /<<PKGBUILDDIR>>/tests/integrationtests/DBusConnectionTests.cpp:35:
/usr/src/googletest/googletest/include/gtest/gtest-printers.h: In instantiation of ‘struct testing::internal::FindFirstPrinter<std::chrono::duration<long int, std::ratio<1, 1000000> >, void, testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter, testing::internal::ProtobufPrinter, testing::internal::ConvertibleToIntegerPrinter, testing::internal::ConvertibleToStringViewPrinter, testing::internal::RawBytesPrinter, testing::internal::FallbackPrinter>’:
```

Would you like to take a look and bump googletest as well?

Hi, Thanks for your efforts and reporting the issue. I'll look at this, I
hope I find time today. Otherwise during the weekend for sure. Kind regards

On Tue, Jan 18, 2022, 08:08 zhsj ***@***.***> wrote:

> Hi, I'm updating sdbus-cpp in Debian
> <https://tracker.debian.org/pkg/sdbus-cpp>. As Debian has bumped
> googletest to 1.11.0, I fail to build sdbus-cpp's test.
>
> Some relevant log
>
> /usr/src/googletest/googletest/include/gtest/gtest-printers.h:291:36: error: no matching function for call to ‘testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter::PrintValue(const std::chrono::duration<long int, std::ratio<1, 1000000> >&, std::nullptr_t)’
>
>   291 |     T, decltype(Printer::PrintValue(std::declval<const T&>(), nullptr)),
>
>       |                 ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>
> /usr/src/googletest/googletest/include/gtest/gtest-printers.h:212:15: note: candidate: ‘template<class T, class, class> static void testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter::PrintValue(const T&, std::ostream*)’
>
>   212 |   static void PrintValue(const T& value, ::std::ostream* os) {
>
>       |               ^~~~~~~~~~
>
> /usr/src/googletest/googletest/include/gtest/gtest-printers.h:212:15: note:   template argument deduction/substitution failed:
>
> /usr/src/googletest/googletest/include/gtest/gtest-printers.h:211:33: error: no match for ‘operator<<’ (operand types are ‘std::basic_ostream<char>’ and ‘const std::chrono::duration<long int, std::ratio<1, 1000000> >’)
>
>   210 |             typename = decltype(std::declval<std::ostream&>()
>
>       |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>
>   211 |                                 << std::declval<const T&>())>
>
>       |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
>
> In file included from /<<PKGBUILDDIR>>/tests/integrationtests/Defs.h:32,
>
>                  from /<<PKGBUILDDIR>>/tests/integrationtests/DBusConnectionTests.cpp:28:
>
>
>
>
> /usr/include/c++/11/ostream:750:5: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
>
> In file included from /usr/src/googletest/googletest/include/gtest/gtest.h:55,
>
>                  from /<<PKGBUILDDIR>>/tests/integrationtests/DBusConnectionTests.cpp:35:
>
> /usr/src/googletest/googletest/include/gtest/gtest-printers.h: In instantiation of ‘struct testing::internal::FindFirstPrinter<std::chrono::duration<long int, std::ratio<1, 1000000> >, void, testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter, testing::internal::ProtobufPrinter, testing::internal::ConvertibleToIntegerPrinter, testing::internal::ConvertibleToStringViewPrinter, testing::internal::RawBytesPrinter, testing::internal::FallbackPrinter>’:
>
>
> Would you like to take a look and bump googletest as well?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Kistler-Group/sdbus-cpp/issues/234>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABM3OUUK5LZ3ZKPA56S2SODUWUGVVANCNFSM5MGL4WYA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -19,7 +19,7 @@ if (NOT TARGET GTest::gmock)
     if (NOT TARGET GTest::gmock)
         include(FetchContent)
 
-        message("Fetching googletest...")
+        message("Fetching googletest v${GOOGLETEST_VERSION}...")
         FetchContent_Declare(googletest
                             GIT_REPOSITORY ${GOOGLETEST_GIT_REPO}
                             GIT_TAG        release-${GOOGLETEST_VERSION}

--- a/tests/integrationtests/Defs.h
+++ b/tests/integrationtests/Defs.h
@@ -56,20 +56,18 @@ const bool DEFAULT_BLOCKING_VALUE{true};
 
 constexpr const double DOUBLE_VALUE{3.24L};
 
-/** Duration stream operator for human readable gtest value output.
- *
- * Note that the conversion to double is lossy if the input type has 64 or more bits.
- * This is ok for our integration tests because they don't have very
- * accurate timing requirements.
- *
- * @return human readable duration in seconds
- */
-template< class Rep, class Period >
-static std::ostream& operator<<(std::ostream& os, const std::chrono::duration<Rep, Period>& d)
-{
-    auto seconds = std::chrono::duration_cast<std::chrono::duration<double>>(d);
-    return os << seconds.count() << " s";
-}
 }}
+
+namespace testing::internal {
+
+// Printer for std::chrono::duration types.
+// This is a workaround, since it's not a good thing to add this to std namespace.
+template< class Rep, class Period >
+void PrintTo(const ::std::chrono::duration<Rep, Period>& d, ::std::ostream* os) {
+    auto seconds = std::chrono::duration_cast<std::chrono::duration<double>>(d);
+    *os << seconds.count() << "s";
+}
+
+}
 
 #endif /* SDBUS_CPP_INTEGRATIONTESTS_DEFS_H_ */

